### PR TITLE
Wow portal fixeroo

### LINKED ISSUE #643: Portal culling's a little buggy
There's a couple instances where portal culling isn't quite working:

* when the camera's very very close to a portal
* looking inside Ironforge's main lava-filled atrium towards the archeologist zone, the museum area gets culled when it shouldn't

### PR DESCRIPTION
**Pull request by wgreenberg**
Fixes #643 and makes portal culling much more effective in general.

### CODE REVIEW

File: src/Geometry.ts [hunk 1]
```diff
@@ -331,32 +331,54 @@ export enum IntersectionState {
...
-    // Left, Right, Top, Bottom, Near, Far
-    public planes: Plane[] = nArray(6, () => new Plane());
+    public planes: Plane[] = [];
+    public near: Plane;
+    public far: Plane;
+
+    constructor(nPlanes: number) {
+        this.planes = nArray(nPlanes, () => new Plane());
+        this.near = new Plane();
+        this.far = new Plane();
```
Comment (magcius) [inline src/Geometry.ts:341]: Why are the near/far planes divorced from the others? Just make them part of the plane array

File: src/Geometry.ts [hunk 2]
```diff
@@ -331,32 +331,54 @@ export enum IntersectionState {
 }
 
 export class Frustum {
-    // Left, Right, Top, Bottom, Near, Far
-    public planes: Plane[] = nArray(6, () => new Plane());
+    public planes: Plane[] = [];
+    public near: Plane;
+    public far: Plane;
+
+    constructor(nPlanes: number) {
```
Comment (magcius) [inline src/Geometry.ts:338]: Mind defaulting this?

```suggestion
    constructor(nPlanes = 6) {
```

File: src/Geometry.ts [hunk 3]
```diff
@@ -331,32 +331,54 @@ export enum IntersectionState {
...
+    }
+
+    public allPlanes(): Plane[] {
+        return this.planes.concat([this.near, this.far]);
+    }
+
+    public copy(o: Frustum) {
+        if (this.planes.length < o.planes.length) {
+            this.planes = nArray(o.planes.length, () => new Plane());
+        }
```
Comment (magcius) [inline src/Geometry.ts:370]:
```suggestion
        while (this.planes.length < o.planes.length)
            this.planes.push(new Plane());
```

File: src/Geometry.ts [hunk 4]
```diff
@@ -331,32 +331,54 @@ export enum IntersectionState {
...
     public updateClipFrustum(m: ReadonlyMat4, clipSpaceNearZ: GfxClipSpaceNearZ): void {
         // http://www8.cs.umu.se/kurser/5DV051/HT12/lab/plane_extraction.pdf
         // Note that we look down the -Z axis, rather than the +Z axis, so we have to invert all of our planes...
+        assert(this.planes.length === 4, `Frustum.updateClipFrustum expects 4 hull planes, this contains ${this.planes.length}`);
 
         this.planes[0].set4Unnormalized(-(m[3] + m[0]), -(m[7] + m[4]), -(m[11] + m[8]) , -(m[15] + m[12])); // Left
-        this.planes[1].set4Unnormalized(-(m[3] - m[0]), -(m[7] - m[4]), -(m[11] - m[8]) , -(m[15] - m[12])); // Right
-        this.planes[2].set4Unnormalized(-(m[3] + m[1]), -(m[7] + m[5]), -(m[11] + m[9]) , -(m[15] + m[13])); // Top
+        this.planes[1].set4Unnormalized(-(m[3] + m[1]), -(m[7] + m[5]), -(m[11] + m[9]) , -(m[15] + m[13])); // Top
+        this.planes[2].set4Unnormalized(-(m[3] - m[0]), -(m[7] - m[4]), -(m[11] - m[8]) , -(m[15] - m[12])); // Right
```
Comment (magcius) [inline src/Geometry.ts:351]: Were these switched for any reason?

**Review by magcius — CHANGES_REQUESTED** [review]

File: src/Geometry.ts (hunk 1 above)
Comment (wgreenberg) [reply src/Geometry.ts:341]: ah yeah, these were from when i was creating a new frustum and copying over the near/far planes. they can be folded in

File: src/Geometry.ts (hunk 4 above)
Comment (wgreenberg) [reply src/Geometry.ts:351]: ah yeah, my frustum viewer debug code needed the planes to be in "rotational" order

**Pull request merged**